keps/sig-scheduling/4815-dra-partitionable-devices/README.md

```diff
@@ -549,105 +802,95 @@ devices:
-pull from it whenever it decides to allocate a "sink" device.
-
-So long as no other devices have been allocated that reference a given "source"
-device in their `ConsumesCapacityField`, it is free to be allocated by the
-scheduler. However, as soon as its capacity has been pulled down by any given
-"sink" device, it can no longer be allocated until its capacity is freed again.
-
-Likewise, so long as all of the advertised capacity of a "sink" device can be
-satisfied by the set of "source" devices it references in its
-`ConsumesCapacityFrom` field, it is free to be allocated by the scheduler.
-However, if any of its advertised capacity cannot be satisfied by one of its
-referenced "source" devices, then it cannot be allocated until that capacity is
-freed by some other device.
-
-Note that in order to support nested partitioning, a "sink" device *may*
-provide a reference to another "sink" device in its `ConsumesCapacityFrom`
-field, so long as:
-
-1. Each device along the recursive chain of references is able to pull enough
-   capacity from the devices in its own `ConsumesCapacityFrom` list to satisfy
-   its allocation.
-
-1. The final device in the chain is a "source" device.
-
-When such a device is allocated, the scheduler will need to track the full
-capacity required to satisfy each of the sink devices along the chain. In this
-way, all intermediate sink devices will essentially be rendered
-"unschedulable", with the last-level sink device pulling its capacity from the
-devices it references directly.
+"gpu-0-partion-2", and "gpu-0-partion-3" respectively. The physical GPU is
+represented as a shared capacity pool, from which all five devices draw their
+capacity. It also sets up a device mixin for the partitions that defines the
+available capacity of the device (available to users) and the consumed capacity
+(how much it draws from the capacity pool). This allows the definition of each
+partition to be quite compact.
+
+After all mixins have been "flattened", the scheduler end up with a list
+of devices that reference a set of pools. It must track the available capacity
+in all capacity pools as devices gets allocated and deallocated, making sure that
+the committed capacity never exceeds what is available.
+
+Allocated resources are tracked in the ResourceClaims and that is where the
+scheduler currently gets information about which devices are available. Using
+this information, the scheduler can compute the available capacity for each
+resource pool. Since each device specifically defines the capacity it needs,
+there is no ambiguity about what capacity is being used by an allocated device.
+
+When the scheduler tries to allocate devices for a new ResourceClaim, it will
+go through the list of devices and check which of them are available. Since it
+knows the free capacity in all capacity pools, it can easily check whether a
+specific device can be allocated.
```

Comment:
Add something like:

Note that since the scheduler will go through these in order and select the *first* (not *best*) fit, it is recommended that `ResourceSlice`s include devices in smallest-to-largest order. Otherwise, the largest one will always be allocated. This may be improved through scoring mechanisms, but that is outside of the scope of this KEP.

And maybe link to the scoring issues.

Reply:
(As an aside, I suppose in theory we could add something in the claim that let the user specify the sort order to use here, rather than just taking what is in the slice. But that sounds...hard and confusing.)